### PR TITLE
[WIP] Feature: create blocks for homepage from dashboard

### PR DESCRIPTION
**Pull request by NyanKiyoshi**
Hello,

This is a feature that was discussed before with a Saleor member. This feature may not be useful for every Saleor users, as they may want edit their home page manually by editing the template `home.html`.

But in some cases, like when we are working with people that don't know anything about HTML and especially much about computers, and want to customize the front page, it's really needed.

The mechanism is rather simple:
- we have a home page and blocks. Like we have right now;
- user creates, edits, deletes, reorder blocks in dashboard;
- user can set:
   - the title (`<h2>`), 
   - the subtitle (`<h3>`),
   - the cover image,
   - the primary button text,
   - the target url (category, collection, page),
   - the disposition (the column size for large, medium and small screen).

----

**Important note:** the reason behind the design change of the home page, is, as the buttons were under the text (subtitle/ title), in the middle of the cards, was making them misaligned between cards which was quite weird looking. So the idea was to put them at the bottom of each card, so we keep a consistency.

----

### Todo list
I'm not done yet, but I open this PR to get thoughts and hopefully to improve the outcome of this feature and the user friendliness and extendability of Saleor.

- [x] Be able to reorder blocks using Ajax (like the menu list page) and drop the position field in the form as it's only there temporally;
- [ ] Be able to set the column size by browsing through the viewports (large, medium, small screen) and selecting a column size for this given viewport (and thus, drop the ugly and non-user friendly `html classes` field);
- [x] Integrate the warmer the current Saleor version;
- [ ] ~~Add localization to the models (to every models through another PR);~~ **Edit: after some thinking and rethinking, I will be adding this in another PR. I'm really looking forward into opening a PR for models localization around this summer, but can't really promise that I have time to work on it (I'll keep Saleor updated if get to implement this feature on my fork).**
- [ ] Random default data should be generated using i18n translation, thus by using the default set language for now, same translations as the hardcoded ones that were in `home.html` as this is almost the exact same thing (only added thing: primary button texts);
- [x] Drop the CSS `:hover` on cover image input field for a Javascript code that activates on hover and on file drag and drop;
- [x] Drop the CSS zoom effect;
- [x] Replace the `<h1>` tag on homepage by `<h2>` as having multiple `<h1>` tags on a single page is invalid;
- [ ] Random data must generate random (or not) links to whether, collections or pages;
- [x] Add a migration generating the blocks data from random_data;
- [ ] Add documentation in the `architecture` section.

#### Dashboard

(outdated gif)
gif

#### Home page

(outdated screenshot)

### Pull Request Checklist

(Please keep this section. It will make maintainer's life easier.)

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [ ] The code is documented (docstrings, project documentation).
1. [x] Python code quality checks pass: `pycodestyle`, `pydocstyle`, `pylint`.
1. [x] JavaScript code quality checks pass: `eslint`.

### CODE REVIEW

File: templates/dashboard/includes/_menu_items.html [hunk 1]
```diff
@@ -32,9 +32,12 @@
           {% endif %}
         </tr>
       </thead>
-      <tbody id="menu-items" class="menu-items" data-post-url="{% if menu_item %}{% url 'dashboard:menu-items-reorder' menu_pk=menu.pk root_pk=menu_item.pk %}{% else %}{% url 'dashboard:menu-items-reorder' menu_pk=menu.pk %}{% endif %}">
+      <tbody class="sortable-items-container"
+             data-post-parameter="ordered_menu_items"
+             data-post-url="{% if menu_item %}{% url 'dashboard:menu-items-reorder' menu_pk=menu.pk root_pk=menu_item.pk %}{% else %}{% url 'dashboard:menu-items-reorder' menu_pk=menu.pk %}{% endif %}"
+      >
```
Comment (NyanKiyoshi) [inline templates/dashboard/includes/_menu_items.html]: I'm not sure about Saleor's code style. Do I have to return `>` to a new line or keep it on the same line?

**Comment by NyanKiyoshi** [comment]
### Latest changes: reposition through dragging

gif

**Comment by NyanKiyoshi** [comment]
On the cards? I had a feedback from a designer that it may cause seasickness to user, so maybe I will get rid of it or reduce the zoom.

**Comment by patrys** [comment]
Yeah, it is vertigo-inducing and no other places in the dashboard use hover animations for cards. I don't think it adds anything to the experience and Material Design guidelines ask for animations to have a guiding purpose.

File: templates/dashboard/includes/_menu_items.html (hunk 1 above)
Comment (patrys) [reply templates/dashboard/includes/_menu_items.html]: Soon™: https://github.com/prettier/prettier/issues/1882

**Comment by NyanKiyoshi** [comment]
As I don't have time to work on this PR, I think I will close it; at least for now.
I'm thinking about coming with something more modular on dashboard 2.0, like Shopify does. 

Be able to edit the home page and manage sections.

Those sections can be:
  - a simple text with a title,
  - a picture with a text and a title,
  - a list of collections (columns),
  - a list of products from a collection (columns),
  - a product to show,
  - a row of columns of images with text overlay (exactly what this PR does),
  - a slideshow.

**Comment by elsagong** [comment]
@NyanKiyoshi Hi, May I have a look of blocks for homepage from dashboard, I can't open these files......

**Comment by NyanKiyoshi** [comment]
@elsagong can't you access the `Files changed` tab? From there you would be able to see the diffs and click `View` and `Raw` to access to full file.

You can also retrieve the whole [diff](https://github.com/mirumee/saleor/pull/2032.diff) and [patch](https://github.com/mirumee/saleor/pull/2032.patch).

Also note that it was last rebased on c9e2d94 (June release).

**Comment by elsagong** [comment]
@NyanKiyoshi Thank you so much, stupid me, finally fount it😁 Do you know what happened with my homepage-blocks of dashboard? Thanks again for your great help🌹
wx20180824-090155 2x

**Comment by NyanKiyoshi** [comment]
@elsagong did you recompile the assets? (`npm run build-assets`)

**Comment by elsagong** [comment]
@NyanKiyoshi Yes, I already recompiled the assets via npm run build-assets, nothing changed 😢

**Comment by elsagong** [comment]
@NyanKiyoshi Hi Dear friend,  page of StoreFront looks  a little messy😂Do you know the reason why they are disordered? Thank you so much!🌹🌹

**Comment by NyanKiyoshi** [comment]
@elsagong it's working fine for me. Maybe you made an error over the bootstrap grid classes? (https://getbootstrap.com/docs/4.1/layout/grid/)

**Comment by elsagong** [comment]
@NyanKiyoshi Hi My dearest friend,  thank you again for replying my message, I am trying to add `carousel photos` in home.hmtl  of homepage-block, may I know why you added a `page_block.html_classes` in home.html? I googled and googled, I didn't find the documents, [link of home.html with `swiper`](https://pastebin.com/pMEgymrK), swiper works, but no matter what I changed, it was always unsatisfactory.

**Comment by NyanKiyoshi** [comment]
@elsagong the `html_classes` were supposed to allow users to customize the blocks, which were columns. For example, `col-sm-12 col-md-6`, which means to make the block full width on mobile screen, and half width on medium screens.

The objective (like said before) was to give the user a button field to allow them to customize how it would render over mobiles, tablets and desktops. The classes were temporary.

This is a field that you can omit for your use case, as if I understand correctly, you want a carousel; which would be full width and only showing one thing at the time.
If so, you should probably fully retake from the ground `_homepage_block.scss` and `templates/home.html` as they were not made for that use case.

**Comment by elsagong** [comment]
@NyanKiyoshi Good morning, got it, thank you so so so much, I will try like you said right now, good day to you and your family💐☕️🍻

**Pull request closed without merging**